File: frontend/types/identity.ts
export type IdentityKind = 'aadhaar' | 'pan';

/**
 * What the app holds about an identity number in the ordinary case.
 *
 * There is no `value` field by design: the plaintext arrives only as the
 * result of an explicit reveal call and is held in component state for a few
 * seconds, never in a store, a cache or the record itself.
 */
export interface IdentityRecord {
  kind: IdentityKind;
  masked: string;
  updatedAt?: string | null;
}

export interface IdentityAuditEntry {
  kind: IdentityKind;
  action: 'created' | 'updated' | 'viewed' | 'deleted' | 'denied';
  success: boolean;
  at: string;
}

File: backend/app/schemas/identity.py
"""Wire shapes for identity numbers.

Note what is absent: no response model here carries the number itself except
`IdentityRevealResponse`, which is returned only by the dedicated reveal
endpoint. The default read model cannot express a plaintext value at all, so a
future change to the router cannot leak one through it by accident.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

IdentityKind = Literal["aadhaar", "pan"]


class IdentityRead(BaseModel):
    """What every listing and write returns: the masked form, never the value."""

    kind: IdentityKind
    masked: str
    updatedAt: datetime | None = None


class IdentityUpsert(BaseModel):
    # Bounded well above any real identifier but far below anything that could
    # be used to smuggle a document through this field.
    value: str = Field(..., min_length=8, max_length=32)


class IdentityRevealResponse(BaseModel):
    """The single shape permitted to carry plaintext, from one endpoint."""

    kind: IdentityKind
    value: str


class IdentityAuditEntry(BaseModel):
    kind: IdentityKind
    action: str
    success: bool
    at: datetime
